File: src/prompt-mode.test.ts
import { describe, expect, it } from "vitest";
import {
  CONVERT_SYSTEM_PROMPTS,
  REVERSE_SYSTEM_PROMPTS,
  SCOPED_REVERSE_SYSTEM_PROMPTS,
} from "./data/prompt-templates";
import {
  buildConvertUserText,
  buildModeRepairUserText,
  buildPromptRuleRepairUserText,
  cleanPromptOutput,
  isLikelyTagListPrompt,
  isLikelyNaturalLanguagePrompt,
  knownCharacterRuntimeInstruction,
  modeNeedsRepair,
  modeUserInstruction,
  modeRepairSystemPrompt,
  promptRuleRepairSystemPrompt,
  promptRuleViolations,
  naturalRepairSystemPrompt,
  parsePromptVariantResponse,
  resolveModePrompt,
} from "./prompt-mode";

describe("resolveModePrompt", () => {
  it("does not let a legacy tag template override convert natural mode", () => {
    const result = resolveModePrompt(
      "natural",
      { tags: "", natural: "", mixed: "" },
      "legacy danbooru tag template",
      CONVERT_SYSTEM_PROMPTS,
    );
    expect(result).toBe(CONVERT_SYSTEM_PROMPTS.natural);
    expect(result).toContain("简洁英文自然语言提示词");
  });

  it("does not let a legacy tag template override reverse natural mode", () => {
    const result = resolveModePrompt(
      "natural",
      { tags: "", natural: "", mixed: "" },
      "legacy reverse tag template",
      REVERSE_SYSTEM_PROMPTS,
    );
    expect(result).toBe(REVERSE_SYSTEM_PROMPTS.natural);
    expect(result).toContain("简洁英文自然语言提示词");
  });

  it("does not let a legacy template override tags mode either", () => {
    const result = resolveModePrompt("tags", { tags: "", natural: "", mixed: "" }, "legacy tags", CONVERT_SYSTEM_PROMPTS);
    expect(result).toBe(CONVERT_SYSTEM_PROMPTS.tags);
  });

  it("uses explicit per-mode overrides before defaults", () => {
    expect(
      resolveModePrompt("natural", { tags: "", natural: "custom natural", mixed: "" }, "legacy tags", CONVERT_SYSTEM_PROMPTS),
    ).toBe("custom natural");
  });
});

describe("prompt mode output handling", () => {
  it("detects the failed natural-mode tag-list shape", () => {
    expect(
      isLikelyTagListPrompt("2boys, black hair, white shirt, sitting, drawing, blue hair, blue hoodie, standing, throwing, ball"),
    ).toBe(true);
  });

  it("does not flag the expected natural multi-character prompt", () => {
    const prompt =
      "Two boys are in a classroom with desks, chairs, a sketchbook, and colored balls, shown from the front in a full-body view | A boy with short black hair and a white shirt is sitting on the left at the desk and drawing in the sketchbook with a pencil | A boy with blue hair and a dark blue hoodie is standing on the right and juggling three colored balls";
    expect(isLikelyTagListPrompt(prompt)).toBe(false);
    expect(isLikelyNaturalLanguagePrompt(prompt)).toBe(true);
  });

  it("cleans common model wrappers without changing prompt semantics", () => {
    expect(cleanPromptOutput("Output: \"Two boys are in a classroom\\n| A boy is drawing\"")).toBe(
      "Two boys are in a classroom | A boy is drawing",
    );
  });

  it("adds hard natural-language constraints to convert requests", () => {
    const text = buildConvertUserText("一个黑发男孩坐着画画", "natural");
    expect(text).toContain("User description:");
    expect(text).toContain("一个黑发男孩坐着画画");
    expect(text).toContain("not a comma-separated tag list");
    expect(text).toContain("base scene | A boy/girl");
  });

  it("builds a repair prompt anchored to the target example style", () => {
    expect(naturalRepairSystemPrompt()).toContain("Two boys are in a classroom");
    expect(buildModeRepairUserText("natural", "一个黑发男孩", "1boy, black hair, sitting")).toContain("Incorrect output:");
  });

  it("repairs tags mode when the model returns pure prose", () => {
    const prose = "Two boys are in a classroom while one boy is drawing and another boy is juggling balls.";
    expect(modeNeedsRepair("tags", prose)).toBe(true);
    expect(modeRepairSystemPrompt("tags")).toContain("comma-separated tags");
  });

  it("does not repair tags mode when the model returns tag-style output", () => {
    const tags =
      "2boys, classroom, desks, chairs, sketchbook, colored balls, full body, from front | boy, short black hair, white shirt, sitting, drawing | boy, blue hair, dark blue hoodie, standing, juggling balls";
    expect(modeNeedsRepair("tags", tags)).toBe(false);
  });

  it("repairs mixed mode when the model returns pure prose only", () => {
    const prose = "Two boys are in a classroom while one boy is drawing and another boy is juggling balls.";
    expect(modeNeedsRepair("mixed", prose)).toBe(true);
    expect(modeRepairSystemPrompt("mixed")).toContain("mostly Danbooru tags");
  });

  it("parses known-character JSON variants", () => {
    const parsed = parsePromptVariantResponse(
      JSON.stringify({
        namePrompt: "1girl, solo, furina (genshin impact), drinking tea",
        featurePrompt: "1girl, solo, white hair, blue eyes, blue outfit, drinking tea",
      }),
      true,
    );
    expect(parsed.primary).toBe("1girl, solo, furina (genshin impact), drinking tea");
    expect(parsed.variants?.featurePrompt).toContain("white hair");
  });

  it("parses labeled known-character variants", () => {
    const parsed = parsePromptVariantResponse(
      "角色名版：1girl, solo, furina (genshin impact)\n特征版：1girl, solo, white hair, blue eyes, blue outfit",
      true,
    );
    expect(parsed.variants?.namePrompt).toContain("furina (genshin impact)");
    expect(parsed.variants?.featurePrompt).not.toContain("furina");
  });

  it("adds concise no-name guidance when known character mode is off", () => {
    const instruction = knownCharacterRuntimeInstruction("tags", "convert", false);
    expect(instruction).toContain("已知网络/游戏/动漫角色模式已关闭");
    expect(instruction).toContain("不要使用角色名");
    expect(instruction).toContain("最少必要");
  });

  it("requires both known-character variants to keep full template detail", () => {
    const instruction = knownCharacterRuntimeInstruction("tags", "convert", true);
    expect(instruction).toContain("namePrompt 和 featurePrompt");
    expect(instruction).toContain("同一完整画面");
    expect(instruction).not.toContain("Keep both prompts short");
    expect(instruction).not.toContain("Only add outfit, feature, pose, action");
  });

  it("adds mature-tag priority only to tags and mixed runtime rules", () => {
    expect(modeUserInstruction("tags", "convert")).toContain("exact mature");
    expect(modeUserInstruction("mixed", "reverse")).toContain("mature tags first");
    expect(modeUserInstruction("natural", "convert")).not.toContain("mature tags first");
    expect(CONVERT_SYSTEM_PROMPTS.natural).not.toContain("成熟整词优先");
  });

  it("detects duplicate, conflicting and decomposed mature tags", () => {
    const issues = promptRuleViolations(
      "tags",
      "1girl, cowboy shot, upper body, smile, smile",
      ["cowboy_shot"],
    );
    expect(issues.some((issue) => issue.includes("重复 Tag：smile"))).toBe(true);
    expect(issues.some((issue) => issue.includes("cowboy shot / upper body"))).toBe(true);
    expect(issues.some((issue) => issue.includes("成熟 Tag cowboy shot"))).toBe(true);
  });

  it("keeps natural mode outside mature-tag validation and builds focused repair", () => {
    expect(
      promptRuleViolations("natural", "A girl is standing.", ["standing"]),
    ).toEqual([]);
    expect(promptRuleRepairSystemPrompt("tags", false)).toContain(
      "只修复明确列出的违规项",
    );
    expect(
      buildPromptRuleRepairUserText({
        mode: "tags",
        originalInput: "七分身女孩",
        draft: "1girl, cowboy shot, upper body",
        violations: ["互斥 Tag"],
        matureTags: ["cowboy_shot"],
      }),
    ).toContain("cowboy_shot");
  });
});

describe("concise NovelAI V5 production templates", () => {
  const six = [
    ...Object.values(SCOPED_REVERSE_SYSTEM_PROMPTS),
    ...Object.values(CONVERT_SYSTEM_PROMPTS),
  ];

  it("keeps all six V5 templates bounded and free of obsolete workflow commands", () => {
    expect(six).toHaveLength(6);
    for (const template of six) {
      expect(template).toContain("NovelAI V5");
      expect(template.length).toBeGreaterThan(1_000);
      expect(template.length).toBeLessThan(2_500);
      expect(template).not.toContain("优先使用 mcp 服务搜索");
      expect(template).not.toContain("不要默认全部无权重");
      expect(template).not.toContain("图片分析顺序");
    }
  });

  it("preserves official dataset, text and multi-character boundaries", () => {
    for (const template of six) {
      expect(template).toContain("fur dataset");
      expect(template).toContain("background dataset");
      expect(template).toContain("Text:");
      expect(template).toContain("最多 22");
      expect(template).toContain("transparent background");
    }
    expect(SCOPED_REVERSE_SYSTEM_PROMPTS.mixed).toContain(
      "角色残差紧跟被限定的 Tag 或动作",
    );
    expect(CONVERT_SYSTEM_PROMPTS.mixed).toContain("自然语言不是必填");
  });

  it("locks in the audited V5 output-quality safeguards", () => {
    for (const template of [
      SCOPED_REVERSE_SYSTEM_PROMPTS.tags,
      SCOPED_REVERSE_SYSTEM_PROMPTS.mixed,
      CONVERT_SYSTEM_PROMPTS.tags,
      CONVERT_SYSTEM_PROMPTS.mixed,
    ]) {
      expect(template).toContain("不得留下孤立锚点");
      expect(template).toContain("1.2::tag ::");
      expect(template).toContain("source#giving/target#giving");
      expect(template).not.toContain("source#handing item");
      expect(template).toContain("交接中的道具不算共享道具");
      expect(template).toContain("属于关键互动");
    }
    for (const template of [
      SCOPED_REVERSE_SYSTEM_PROMPTS.natural,
      CONVERT_SYSTEM_PROMPTS.natural,
    ]) {
      expect(template).toContain("text, <language> text");
      expect(template).toContain("不复述文字内容");
      expect(template).not.toContain("reads OPEN");
    }
    for (const template of six) {
      expect(template).not.toContain("base 最末、第一个 | 之前");
      expect(template).not.toContain("不写 portrait、landscape");
      expect(template).toContain("同一层级互斥");
      expect(template).toContain("不视为互斥");
    }
    expect(SCOPED_REVERSE_SYSTEM_PROMPTS.tags).toContain(
      "本模式允许省略且不得混入自然语言",
    );
    expect(CONVERT_SYSTEM_PROMPTS.tags).toContain(
      "空间关系优先由角色段顺序表达",
    );
    expect(SCOPED_REVERSE_SYSTEM_PROMPTS.natural).not.toContain(
      "人数 Tag/人数描述",
    );
    expect(CONVERT_SYSTEM_PROMPTS.natural).not.toContain(
      "人数 Tag/人数描述",
    );
    expect(CONVERT_SYSTEM_PROMPTS.tags).toContain("mutual#holding hands");
    expect(CONVERT_SYSTEM_PROMPTS.mixed).toContain("mutual#holding hands");
    expect(SCOPED_REVERSE_SYSTEM_PROMPTS.mixed).toContain(
      "无成熟 Tag 的关键可见状态或表情",
    );
    expect(CONVERT_SYSTEM_PROMPTS.mixed).toContain(
      "无成熟 Tag 的关键可见状态或表情",
    );
    for (const template of Object.values(CONVERT_SYSTEM_PROMPTS)) {
      expect(template).toContain("-1::");
      expect(template).toContain("Text:");
    }
  });
});

File: mobile/test/prompt_mode_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:novelai_mobile/models/nai_models.dart';
import 'package:novelai_mobile/prompts/prompt_mode.dart';
import 'package:novelai_mobile/prompts/prompt_templates.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  test('normal mode returns one cleaned prompt without variants', () {
    final result = parsePromptVariantResponse(
      '```text\nPrompt: 1girl,  solo, blue hair\n```',
      false,
    );
    expect(result.primary, '1girl, solo, blue hair');
    expect(result.variants, isNull);
  });

  test('known character mode parses strict JSON variants', () {
    final result = parsePromptVariantResponse(
      '{"namePrompt":"1girl, solo, furina (genshin impact)",'
      '"featurePrompt":"1girl, solo, white hair, blue eyes, blue coat"}',
      true,
    );
    expect(result.primary, contains('furina'));
    expect(result.variants?.namePrompt, contains('furina (genshin impact)'));
    expect(result.variants?.featurePrompt, isNot(contains('furina')));
    expect(result.variants?.isComplete, isTrue);
  });

  test('known character mode accepts labeled fallback output', () {
    final result = parsePromptVariantResponse(
      '角色名版：1girl, solo, furina (genshin impact)\n'
      '特征版：1girl, solo, white hair, blue eyes, blue coat',
      true,
    );
    expect(result.variants?.isComplete, isTrue);
  });

  test('runtime rule keeps known and unknown character behavior distinct', () {
    final known = knownCharacterRuntimeInstruction(
      ReversePromptMode.tags,
      'convert',
      true,
    );
    final unknown = knownCharacterRuntimeInstruction(
      ReversePromptMode.tags,
      'convert',
      false,
    );
    expect(known, contains('namePrompt 和 featurePrompt'));
    expect(known, contains('同一完整画面'));
    expect(known, contains('规范 Danbooru 角色 Tag'));
    expect(known, contains('高置信度标志性外貌、默认服装与配饰'));
    expect(known, contains('个人法典规则'));
    expect(known, isNot(contains('Keep both prompts short')));
    expect(unknown, contains('不要使用角色名'));
  });

  test('desktop templates are bundled for all modes', () async {
    final library = await PromptTemplateLibrary.load();
    for (final mode in ReversePromptMode.values) {
      expect(library.get('reverse', mode).length, greaterThan(500));
      expect(library.get('convert', mode).length, greaterThan(500));
      expect(library.get('scopedReverse', mode).length, greaterThan(300));
      expect(
          library
              .getReverse(mode, scoped: false, templateVersion: 'v4.5')
              .length,
          greaterThan(300));
      expect(library.get('comic', mode).length, greaterThan(300));
    }
  });

  test('custom prompt template settings survive JSON persistence', () {
    final settings = AppSettings(
      reversePromptTemplates: {'tags': 'custom reverse'},
      convertPromptTemplates: {'mixed': 'custom convert'},
      comicPromptTemplate: 'custom comic',
      promptRuleAutoRepairEnabled: true,
      reversePromptTemplateVersion: 'v4.5',
    );
    final restored = AppSettings.fromJson(settings.toJson());
    expect(restored.reversePromptTemplates['tags'], 'custom reverse');
    expect(restored.convertPromptTemplates['mixed'], 'custom convert');
    expect(restored.comicPromptTemplate, 'custom comic');
    expect(restored.promptRuleAutoRepairEnabled, isTrue);
    expect(restored.reversePromptTemplateVersion, 'v4.5');
  });

  test('mature tag rules apply only to tags and mixed modes', () {
    expect(modeUserInstruction(ReversePromptMode.tags, 'convert'),
        contains('exact mature'));
    expect(modeUserInstruction(ReversePromptMode.mixed, 'reverse'),
        contains('mature Danbooru'));
    expect(modeUserInstruction(ReversePromptMode.natural, 'convert'),
        isNot(contains('mature Danbooru')));
  });

  test('known-character conversion uses a strict paired JSON contract', () {
    final instruction = modeUserInstruction(
      ReversePromptMode.mixed,
      'convert',
      knownCharacter: true,
    );
    expect(instruction, contains('exactly two string fields'));
    expect(instruction, contains('namePrompt'));
    expect(instruction, contains('featurePrompt'));
    expect(
        instruction, contains('signature appearance, outfit, and accessories'));
    expect(instruction, contains('75–85%'));
  });

  test('bundled reverse and convert templates use the concise V5 contract',
      () async {
    final library = await PromptTemplateLibrary.load();
    for (final kind in ['scopedReverse', 'convert']) {
      for (final mode in ReversePromptMode.values) {
        final template = library.get(kind, mode);
        expect(template, contains('NovelAI V5'));
        expect(template.length, inInclusiveRange(1000, 2700));
        expect(template, contains('fur dataset'));
        expect(template, contains('background dataset'));
        expect(template, contains('Text:'));
        expect(template, contains('最多 22'));
        expect(template, contains('transparent background'));
        expect(template, isNot(contains('优先使用 mcp 服务搜索')));
        expect(template, isNot(contains('不要默认全部无权重')));
        expect(template, isNot(contains('图片分析顺序')));
      }
    }
  });

  test('bundled templates preserve audited V5 prompt-quality safeguards',
      () async {
    final library = await PromptTemplateLibrary.load();
    for (final kind in ['scopedReverse', 'convert']) {
      for (final mode in [ReversePromptMode.tags, ReversePromptMode.mixed]) {
        final template = library.get(kind, mode);
        expect(template, contains('不得留下孤立锚点'));
        expect(template, contains('1.2::tag ::'));
        expect(template, contains('source#giving/target#giving'));
        expect(template, isNot(contains('source#handing item')));
        expect(template, contains('交接中的道具不算共享道具'));
        expect(template, contains('属于关键互动'));
      }
      final natural = library.get(kind, ReversePromptMode.natural);
      expect(natural, contains('text, <language> text'));
      expect(natural, contains('不复述文字内容'));
      expect(natural, isNot(contains('reads OPEN')));
      for (final mode in ReversePromptMode.values) {
        final template = library.get(kind, mode);
        expect(template, isNot(contains('base 最末、第一个 | 之前')));
        expect(template, isNot(contains('不写 portrait、landscape')));
        expect(template, contains('同一层级互斥'));
        expect(template, contains('不视为互斥'));
      }
    }
    expect(
      library.get('scopedReverse', ReversePromptMode.mixed),
      contains('无成熟 Tag 的关键可见状态或表情'),
    );
    expect(
      library.get('scopedReverse', ReversePromptMode.mixed),
      contains('其他关系短语紧跟被限定的 Tag 或动作'),
    );
    expect(
      library.get('scopedReverse', ReversePromptMode.tags),
      contains('本模式允许省略且不得混入自然语言'),
    );
    expect(
      library.get('convert', ReversePromptMode.tags),
      contains('空间关系优先由角色段顺序表达'),
    );
    expect(
      library.get('convert', ReversePromptMode.tags),
      contains('mutual#holding hands'),
    );
    expect(
      library.get('convert', ReversePromptMode.mixed),
      contains('mutual#holding hands'),
    );
    expect(
      library.get('convert', ReversePromptMode.mixed),
      contains('无成熟 Tag 的关键可见状态或表情'),
    );
    expect(
      library.get('convert', ReversePromptMode.mixed),
      allOf(contains('75–85%'), contains('15–25%'), contains('示例中文含义')),
    );
  });

  test('rule validator detects duplicate and mature tag decomposition', () {
    final issues = promptRuleViolations(
      ReversePromptMode.tags,
      '1girl, cowboy shot, upper body, smile, smile',
      ['cowboy_shot'],
    );
    expect(issues.any((item) => item.contains('重复 Tag：smile')), isTrue);
    expect(issues.any((item) => item.contains('cowboy shot / upper body')),
        isTrue);
    expect(issues.any((item) => item.contains('成熟 Tag cowboy shot')), isTrue);
    expect(
        promptRuleViolations(
            ReversePromptMode.natural, 'A girl is standing.', ['standing']),
        isEmpty);
  });

  test('mixed validator rejects pure tags but accepts tag plus relation prose',
      () {
    const tags =
        '2girls, cafe, indoors, evening, upper body, counter, cake | girl, black hair, green eyes, holding plate | girl, red hair, blue eyes, reaching';
    expect(
      promptRuleViolations(ReversePromptMode.mixed, tags),
      contains('混合模式缺少约 20% 的自然语言关系短语'),
    );
    expect(
      promptRuleViolations(
        ReversePromptMode.mixed,
        '$tags, on the right, reaching with both hands',
      ).where((item) => item.contains('混合模式')),
      isEmpty,
    );
  });
}
